lib/crx.js

```diff
@@ -0,0 +1,164 @@
+    signer.update(CRX_SIGNED_DATA_HEADER)
+    signer.update(toLE32(header.signed_header_data.length))
+    signer.update(header.signed_header_data)
+    signer.update(content)
+
+    const publicKey = crypto.createPublicKey(signingKey)
+    const proof = {
+      public_key: publicKey.export({ type: 'spki', format: 'der' }),
+      signature: signer.sign(signingKey)
+    }
+    // Add signature.
+    header.sha256_with_rsa.push(proof)
+  }
+  return header
+}
+
+const zipContent = async (stagingDir) => {
+  const processFiles = (dir, callback) => {
+    const files = glob.sync('**/*', {
+      nodir: true,
+      absolute: false,
+      cwd: dir
+    })
+
+    for (const relativePath of files) {
+      callback(relativePath)
+    }
+  }
+
+  const zip = new JSZip()
+  processFiles(stagingDir, (file) => {
+    zip.file(file, fs.readFileSync(path.join(stagingDir, file)))
+  })
+
+  return await zip.generateAsync({
+    type: 'Uint8Array',
+    compression: 'DEFLATE',
+    compressionOptions: { level: 9 }
+  })
+}
+
+const updateManifest = (stagingDir, extensionKey) => {
+  const manifestFile = path.join(stagingDir, 'manifest.json')
+  const manifest = util.parseManifest(manifestFile)
+  const publicKey = crypto.createPublicKey(extensionKey)
+
+  // Public key which used to generate the extension/crx id.
+  manifest.key = publicKey
+    .export({
+      type: 'spki',
+      format: 'der'
+    })
+    .toString('base64')
+
+  // The WebStore update URL makes the browser think that the extension is from the Web Store.
+  manifest.update_url = 'https://clients2.google.com/service/update2/crx'
+
+  fs.writeFileSync(manifestFile, JSON.stringify(manifest, null, 2))
+  return manifest
+}
+
+// Generates
+const generateCrx = async (
+  stagingDir,
+  extensionKeyFile,
+  publisherKeyFiles,
+  verifiedContentsKeyFile
+) => {
+  const extensionKey = crypto.createPrivateKey(
+    fs.readFileSync(extensionKeyFile, 'utf-8')
+  )
+  const publisherKeys = publisherKeyFiles
+    .filter((file) => file)
+    .map((file) => {
+      return crypto.createPrivateKey(fs.readFileSync(file))
+    })
+
+  const manifest = updateManifest(stagingDir, extensionKey)
+
+  const zip = await zipContent(stagingDir)
+  const header = signContentAndCreateHeader(zip, extensionKey, publisherKeys)
+  if (verifiedContentsKeyFile) {
+    header.verified_contents = zlib.gzipSync(
+      util.generateVerifiedContents(stagingDir, ['**'], verifiedContentsKeyFile)
+    )
+  }
+
+  const pbf = new Pbf()
+  crxpb.writeCrxFileHeader(header, pbf)
+  const headerData = pbf.finish()
+
+  const crx = Buffer.alloc(
+    CRX_FILE_MAGIC.length + 4 /* LE32 version */ + 4 /* LE32 header length */
+  )
+
+  let pos = crx.write(CRX_FILE_MAGIC, 0, 'utf-8')
+  pos = crx.writeUint32LE(CRX_VERSION, pos)
+  crx.writeUint32LE(headerData.length, pos)
+
+  return {
+    crx: Buffer.concat([crx, headerData, zip]),
+    zip,
+    manifest
+  }
+}
+
+export default {
+  generateCrx
```

Comment:
any way to have a unit test for this? (check that generated CRX signature is as expected and the signature can be verified)

Reply:
Added.